File: Python/LE002AddTwoNumbers.py
from utility.ListNode import ListNode

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        node = dummy
        inc = False
        newval = 0
        while l1 != None or l2 != None or inc :
            newval = 1 if inc else 0
            if l1 != None:
                newval += l1.val
                l1 = l1.next
            if l2 != None:
                newval += l2.val
                l2 = l2.next
            inc = True if newval >= 10 else False
            node.next = ListNode(newval % 10)
            # print(str(node.val))
            node = node.next
        return dummy.next

l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l1.next = l2
l2.next = l3
l3.next = l4
l11 = ListNode(1)
l12 = ListNode(2)
l13 = ListNode(3)
l14 = ListNode(6)
l11.next = l12
l12.next = l13
l13.next = l14


if __name__ == '__main__':
    test = Solution()
    l = test.addTwoNumbers(l1, l11)
    l.print()

File: LeetcodeCpp/LE002addTwoNumbers.cpp
#include <iostream>
#include <string>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2);
};
ListNode* Solution::addTwoNumbers(ListNode* l1, ListNode* l2) {
	ListNode *dummy = new ListNode(0);
	ListNode *node;
	node = dummy;
	int newval;
	bool inc = false;
	while(l1 != NULL || l2 != NULL || inc){
		newval = inc ? 1 : 0;
		if(l1 != NULL){
			newval += l1->val;
			l1 = l1->next;
		}
		if(l2 != NULL){
			newval += l2->val;
			l2 = l2->next;
		}
		inc = newval >= 10 ? true : false;
		node->next = new ListNode(newval % 10);
		node = node->next;
	}
	node = NULL;
	return dummy->next;
}

void printList(ListNode* head){
	while(head != NULL){
		cout << head->val << " ";
		head = head->next;
	}
	cout << endl;
}
int main(){
	ListNode *l1 = new ListNode(1);
	ListNode *l2 = new ListNode(2);
	ListNode *l3 = new ListNode(3);
	ListNode *l4 = new ListNode(6);
	ListNode *l5 = new ListNode(9);
	l1->next = l2;
	l2->next = l3;
	l3->next = l4;
	l4->next = l5;
	ListNode *l11 = new ListNode(1);
	ListNode *l12 = new ListNode(2);
	ListNode *l13 = new ListNode(3);
	ListNode *l14 = new ListNode(4);
	l11->next = l12;
	l12->next = l13;
	l13->next = l14;
	// printList(l1);
	Solution sol;
	ListNode *test = sol.addTwoNumbers(l1, l11);
	printList(test);
	return 0;
}